ports/nrf/common-hal/nvm/ByteArray.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Nick Moore for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "common-hal/nvm/ByteArray.h"
+
+#include <stdio.h>
+#include <string.h>
+
+#include "nrf_nvmc.h"
+
+uint32_t common_hal_nvm_bytearray_get_length(nvm_bytearray_obj_t *self) {
+    return self->len;
+}
+
+static void write_page(uint32_t page_addr, uint32_t offset, uint32_t len, uint8_t *bytes) {
+    // Write a whole page to flash, buffering it first and then erasing and rewriting 
+    // it since we can only clear a whole page at a time.
+    // TODO (maybe) check if only clearing bits (don't need to erase)
+    // XXX should this suspend interrupts so erase/write is atomic?
+
+    if (offset == 0 && len == FLASH_PAGE_SIZE) {
+        nrf_nvmc_page_erase(page_addr);
```

Comment:
It's not safe to use the the `nrf_nvmc...` operations when the softdevice is in use. You need to use the `sd_flash_...` flash operations instead. See `ports/nrf/supervisor/internal_flash.c`, which we use when we're using an internal filesystem instead of the QSPI flash chip.

It may be worth generalizing the API in `internal_flash.c` so it can be used for both `nvm` and internal filesystem operations, perhaps factoring out the basic write operation into a callable routine. Right now the `supervisor_flash_flush()` operations assumes it's writing `_flash_cache`.

Reply:
I realize using `internal_flash.c` itself is not the right way to go, because that's conditionally compiled, and included only when there's an internal filesystem. But a low-level flash-writing routine could be factored out to nrf/peripherals or something like that. May or may not be worth it.

Reply:
No worries, that makes sense & shouldn't be too hard to get going.